slides/advanced-01-tuning-hyperparameters.qmd

```diff
@@ -0,0 +1,697 @@
+```
+
+## Choose a parameter combination `r hexes(c("tune"))`
+
+Create your own tibble for final parameters or use one of the `tune::select_*()` functions:
+
+```{r}
+lgbm_best <- select_best(lgbm_res, metric = "mae")
+lgbm_best
+```
+
+## Checking Calibration `r hexes(c("tune", "probably"))`
+
+```{r}
+#| label: lgb-cal-plot
+#| output-location: column
+#| out-width: 90%
+#| fig-width: 5
+#| fig-height: 5
+
+library(probably)
+lgbm_res %>%
+  collect_predictions(
+    parameters = lgbm_best
+  ) %>%
+  cal_plot_regression(
+    truth = avg_price_per_room,
+    estimate = .pred,
+    alpha = 1 / 3
+  )
+```
+
+## Updating the Workflow
+
+If we like this model, we might want to splice the best tuning parameter values into our `lgbm_wflow` object. 
+
+There is a function for that: 
+
+```{r}
+#| label: lgb-finalize
+
+lgbm_wflow <- lgbm_wflow %>% finalize_workflow(lgbm_best)
+lgbm_wflow
+```
+
+
+## Running in parallel
+
+::: columns
+::: {.column width="60%"}
+-   Grid search, combined with resampling, requires fitting a lot of models!
+
+-   These models don't depend on one another and can be run in parallel.
+
+We can use a *parallel backend* to do this:
+
+```{r, eval= FALSE}
+#| eval: false
+#| label: parallel-methods
+cores <- parallelly::availableCores(logical = FALSE)
+cl <- parallel::makePSOCKcluster(cores)
+doParallel::registerDoParallel(cl)
+
+# Now call `tune_grid()`!
+
+# Shut it down with:
+foreach::registerDoSEQ()
+parallel::stopCluster(cl)
+```
+:::
+
+::: {.column width="40%"}
+```{r}
+#| label: resample-times
+#| echo: false
+#| out-width: '100%'
+#| fig-width: 6
+#| fig-height: 6
+#| fig-align: 'center'
+#| dev-args: list(bg = "transparent")
+load("resamples_times.RData")
+resamples_times %>%
+  dplyr::rename(operation = label) %>% 
+  ggplot(aes(y = id_alt, x = duration, fill = operation)) +
+  geom_bar(stat = "identity", color = "black") +
+  labs(y = NULL, x = "Elapsed Time") + 
+  scale_fill_brewer(palette = "Paired") +
+  theme(legend.position = "top")
+```
+:::
+:::
+
+## Running in parallel
+
+Speed-ups are fairly linear up to the number of physical cores (10 here).
+
+```{r}
+#| label: parallel-speedup
+#| echo: false
+#| out-width: '90%'
+#| fig-width: 9
+#| fig-height: 4
+#| fig-align: 'center'
+#| dev-args: list(bg = "transparent")
+load("xgb_times.RData")
+ggplot(times, aes(x = num_cores, y = speed_up, color = parallel_over, shape = parallel_over)) + 
+  geom_abline(lty = 1) + 
+  geom_point(size = 2) + 
+  geom_line() +
+  facet_wrap(~ preprocessing) + 
+  coord_obs_pred() + 
+  scale_color_manual(values = c("#7FC97F", "#386CB0")) +
+  labs(x = "Number of Workers", y = "Speed-up")  +
+  theme(legend.position = "top")
+```
+
+:::notes
+Faceted on the expensiveness of preprocessing used.
+:::
+
+## Your turn {transition="slide-in"}
+
+![](images/parsnip-flagger.jpg){.absolute top="0" right="0" width="150" height="150"}
+
+<br>
+
+What if we used early stopping instead of optimizing the number of trees? 
```

Comment:
This feels somewhat ambiguous to me but is probably fine with some spoken clarification. We mean(?):

1) Can you describe the procedural change in how this model is fitted?
2) Would we expect tuning to take longer? Shorter? The same amount of time?
3) Based on the graphs shown before, do we have any expectations about the performance of these candidates as compared to the previous ones?

Reply:
I'll talk about a lot of this when we get to the slide, but for clarity: 

1. We keep training to a specific maximum number of trees and stop when the validation statistics are consistently worse. 
2. Should take much less time. 
3. Performance should be about the same, perhaps slightly worse (like 1%-2% worse). 

Since the learning rate dominates everything, an argument can be made that the number of trees isn't that consequential. There is probably an interaction between the learning rate and number of trees though.